File: 698-partition-to-k-equal-sum-subsets/698-partition-to-k-equal-sum-subsets.go
func canPartitionKSubsets(nums []int, k int) bool {
  if k == 1 { return true }
  n := len(nums)
  if k > n {
    return false
  }
  sum := 0
  sort.Ints(nums)
  for _, v := range nums {
    sum += v
  }
  if sum % k != 0 { return false }
  mean := sum / k
  if nums[n-1] > mean { return false }
  used := 0
  memo := make(map[int]bool)
  var backtrack func(l, start, t int)bool
  backtrack = func(l, start, t int)bool {
    if l == k { return true }
    if t == 0 {
      res := backtrack(l+1, 0, mean)
      memo[used] = res
      return res
    }
    if res, ok := memo[used]; ok {
      return res
    }
    pre := 0
    for j := start; j < n; j++ {
      if used >> j & 1 == 1 { continue }
      v := nums[j]
      if t < v { return false }
      if v == pre { continue }
      used |= 1 << j
      pre = v
      if backtrack(l, j + 1, t - v) {
        return true
      }
      used &= ^(1 << j)
    }
    return false
  }
  return backtrack(0, 0, mean)
}

File: 698-partition-to-k-equal-sum-subsets/698-partition-to-k-equal-sum-subsets.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {boolean}
 */
var canPartitionKSubsets = function(nums, k) {
  const n = nums.length
  if (k > n) { return false }
  const sum = nums.reduce((a, b) => a+b, 0)
  if (sum % k != 0) { return false }
  const mean = sum / k
  nums.sort((a, b) => a - b);
  if (nums.at(-1) > mean) { return false }
  let used = 0
  const map = {}
  const backtrack = (bucket, start, target) => {
    if (bucket === k) return true
    if (target === 0) {
      const res = backtrack(bucket+1, 0, mean)
      map[used] = res
      return res
    }
    if (map[used] !== undefined) { return map[used] }
    let pre = 0
    for (let i = start; i < n; i++) {
      if (used & (1 << i)) continue
      const v = nums[i]
      if (v > target) return false
      if (v === pre) continue
      used |= (1 << i)
      pre = v
      const res = backtrack(bucket, i + 1, target - v)
      map[used] = res
      used &= ~(1 << i)
      if (res) return true
    }
    return false
  }
  return backtrack(0, 0, mean)
};